src/models/pull-request.ts

```diff
@@ -0,0 +1,172 @@
+  let graphql = githubGraphqlClient();
+  let data: any = await graphql(singlePullRequest(number));
+  return new PullRequest(data.repository.pullRequest as GraphQLPRData);
+}
+
+function findLast<T>(arr: Array<T>, predicate: (item: T) => boolean): T | null {
+  for (let i = arr.length - 1; i >= 0; i--) {
+    if (predicate(arr[i])) {
+      return arr[i];
+    }
+  }
+  return null;
+}
+
+interface CreationDetails {
+  __typename: string;
+  createdAt: string;
+  author: { login: string };
+}
+
+interface PullRequestReview extends CreationDetails {
+  __typename: 'PullRequestReview';
+  state:
+    | 'PENDING'
+    | 'COMMENTED'
+    | 'APPROVED'
+    | 'CHANGES_REQUESTED'
+    | 'DISMISSED';
+}
+
+interface ReviewRequestedEvent extends CreationDetails {
+  __typename: 'ReviewRequestedEvent';
+}
+
+interface GraphQLPRData {
+  mergedAt: string | null;
+  number: number;
+  merged: boolean;
+  createdAt: string;
+  author: { login: string };
+  timelineItems: {
+    nodes: Array<PullRequestReview | ReviewRequestedEvent>;
+  };
+}
+
+type TimelineItem = {
+  kind:
+    | 'PullRequestReview'
+    | 'ReviewRequestedEvent'
+    | 'ReopenedEvent'
+    | 'ReadyForReviewEvent';
+  datetime: DateTime;
+  login: string;
+  node: any;
+};
+
+export class PullRequest {
+  number: number;
+  mergedAt: DateTime | undefined;
+  createdAt: DateTime;
+  debug: Debugger;
+  constructor(public data: GraphQLPRData) {
+    if (data.mergedAt) {
+      this.mergedAt = DateTime.fromISO(data.mergedAt);
+    }
+    this.number = data.number;
+    this.createdAt = DateTime.fromISO(data.createdAt);
+    this.debug = debug.extend('pull-request:' + this.number);
+  }
+
+  @memoize()
+  get timeToMerge(): Duration | undefined {
+    let debug = this.debug.extend('timeToMerge');
+    if (!this.data.merged) {
+      debug('not merged, returning');
+      return undefined;
+    }
+
+    if (this.openedForReviewAt && this.mergedAt) {
+      let diff = this.mergedAt.diff(this.openedForReviewAt);
+      debug(
+        'return %o (openedForReviewAt: %s -> mergedAt: %s)',
+        diff.toObject(),
+        this.openedForReviewAt,
+        this.mergedAt
+      );
+      return diff;
+    } else {
+      return undefined;
+    }
+  }
+
+  @memoize()
+  get timelineItemsAsc(): Array<TimelineItem> {
+    let events = this.data.timelineItems.nodes.map((node: any) => {
+      let kind = node.__typename;
+      let datetime = DateTime.fromISO(node.createdAt);
+      let login = node.actor?.login || node.author?.login;
+      return { kind, datetime, login, node };
+    });
+    events.sort((a: any, b: any) => a.createdAt - b.createdAt);
+    return events;
+  }
+
+  // The PR is opened for review when:
+  // - only counting events that occur after the *last* "reopened" event (if any):
+  //   - last "ReadyForReview" event occurs (aka convert FROM draft to open)
+  //   - OR: first ReviewRequested
+  @memoize()
+  get openedForReviewAt(): DateTime | void {
```

Comment:
Why `DateTime | void`?

Reply:
This was me learning TS a bit. I wanted to avoid having to have a `return` statement in the function if there is no `openedForReviewAt`, and `void` allowed it. Having `DateTime | undefined` caused tsc to complain if any branch of the function didn't have an explicit `return`.

I've since refactored the function though — it always returns something now (DateTime or undefined), so I'll change this.

I've also learned in the meantime that using `!` to narrow a X-or-undefined type only works for `X | undefined`, not `X | void`, so even more reason to use `undefined` here.

done